### ISSUE TITLE
Assert error in http_fail(), cache/cache_http.c line 120:   Condition(WS_Overflowed(hp->ws)) not true.

### ISSUE THREAD

**Issue by simonvik** [varnishcache/varnish-cache#3410]
```
Panic at: Tue, 15 Sep 2020 06:21:25 GMT
Assert error in http_fail(), cache/cache_http.c line 120:
  Condition(WS_Overflowed(hp->ws)) not true.
version = varnish-trunk revision 7786bd8d1396f3a60c23057091c5b34b81e3e275, vrt api = 11.0
ident = Linux,4.15.18+20190618.1109.03d21a5f.bionic,x86_64,-junix,-smalloc,-sdefault,-hcritbit,epoll
now = 39288737.337575 (mono), 1600150884.796663 (real)
Backtrace:
  0x559533a5aedf: /usr/sbin/varnishd(+0x51edf) [0x559533a5aedf]
  0x559533acd073: /usr/sbin/varnishd(VAS_Fail+0x43) [0x559533acd073]
  0x559533a502d1: /usr/sbin/varnishd(+0x472d1) [0x559533a502d1]
  0x559533a5e76d: /usr/sbin/varnishd(Resp_Setup_Deliver+0x13d) [0x559533a5e76d]
  0x559533a5f16e: /usr/sbin/varnishd(CNT_Request+0x60e) [0x559533a5f16e]
  0x559533a91bd8: /usr/sbin/varnishd(h2_do_req+0x68) [0x559533a91bd8]
  0x559533a7f917: /usr/sbin/varnishd(+0x76917) [0x559533a7f917]
  0x559533a7fe04: /usr/sbin/varnishd(+0x76e04) [0x559533a7fe04]
  0x7f46dafde6db: /lib/x86_64-linux-gnu/libpthread.so.0(+0x76db) [0x7f46dafde6db]
  0x7f46dad07a3f: /lib/x86_64-linux-gnu/libc.so.6(clone+0x3f) [0x7f46dad07a3f]
pthread.self = 0x7f467f147700
pthread.name = (cache-worker)
pthread.attr = {
  guard = 4096,
  stack_bottom = 0x7f467f118000,
  stack_top = 0x7f467f148000,
  stack_size = 196608,
}
thr.req = 0x7f469ac5b020 {
  vxid = 5570577, transport = H2
  step = R_STP_DELIVER,
  req_body = none,
  restarts = 0, esi_level = 0,
  sp = 0x7f46b6c3a920 {
    fd = 1134, vxid = 5570576,
    t_open = 1600150883.552320,
    t_idle = 1600150883.559246,
    ws = 0x7f46b6c3a960 {
      id = \"ses\",
      {s, f, r, e} = {0x7f46b6c3a9b8, +152, (nil), +576},
    },
    transport = H2 {
      streams {
        0x00000000 idle  
        0x00000001 c-rem 
      }
    }
    client = 2a02:2350:5:10::1022 43654 /var/run/varnish.sock,
  },
  worker = 0x7f467f146d70 {
    ws = 0x7f467f146e18 {
      id = \"wrk\",
      {s, f, r, e} = {0x7f467f146360, +0, (nil), +2040},
    },
    VCL::method = MISS,
    VCL::return = fetch,
    VCL::methods = {RECV, HASH, MISS},
  },
  ws = 0x7f469ac5b170 {
    id = \"req\",
    {s, f, r, e} = {0x7f469ac5d0d0, +872, (nil), +122664},
  },
  http_conn = 0x7f469ac5d048 {
    doclose = NULL,
    ws = (nil) {
    },
    {rxbuf_b, rxbuf_e} = {(nil), (nil)},
    {pipeline_b, pipeline_e} = {(nil), (nil)},
    content_length = 0,
    body_status = NULL,
    first_byte_timeout = 0.000000,
    between_bytes_timeout = 0.000000,
  },
  http[req] = 0x7f469ac5b210 {
    ws = 0x7f469ac5b170 {
      [Already dumped, see above]
    },
    hdrs {
      \"GET\",
      \"/product-category/womens/******\",
      \"HTTP/2.0\",
      \"scheme: https\",
      \"user-agent: curl/7.72.0\",
      \"accept: */*\",
      \"X-Forwarded-For: 2a02:2350:5:10::1022\",
      \"X-Onecom-Forwarded-IP: 2a02:2350:5:10::1022\",
      \"X-Onecom-Forwarded-Proto: https\",
      \"X-Forwarded-Proto: https\",
      \"X-Onecom-Host: ******.com\",
      \"X-Debug: true\",
      \"host: ******.com\",
    },
  },
  http[resp] = 0x7f469ac5bb00 {
    ws = 0x7f469ac5b170 {
      [Already dumped, see above]
    },
    hdrs {
      \"HTTP/1.1\",
      \"200\",
      \"OK\",
      \"Date: Tue, 15 Sep 2020 06:21:23 GMT\",
      \"Server: Apache\",
      \"X-Powered-By: PHP/7.2.33\",
      \"Expires: Wed, 11 Jan 1984 05:00:00 GMT\",
      \"Cache-Control: no-transform, no-cache, no-store, must-revalidate\",
      \"Link: <https://******.com/wp-json/>; rel=\"https://api.w.org/\"\",
      \"Set-Cookie: *************; expires=Wed, 16-Sep-2020 06:21:23 GMT; Max-Age=86400; path=/; secure\",
      \"Set-Cookie: *************; expires=Wed, 16-Sep-2020 06:21:23 GMT; Max-Age=86400; path=/; secure\",
      \"Set-Cookie: *************; expires=Wed, 16-Sep-2020 06:21:23 GMT; Max-Age=86400; path=/; secure\",
      \"Set-Cookie: *************; expires=Wed, 16-Sep-2020 06:21:23 GMT; Max-Age=86400; path=/; secure\",
      \"Set-Cookie: alg-wc-wl-user-id=*************; expires=Mon, 14-Dec-2020 06:21:23 GMT; Max-Age=7776000; path=/; secure\",
      \"Set-Cookie: woocommerce_items_in_cart=1; path=/\",
      \"Set-Cookie: woocommerce_cart_hash=*************; path=/\",
      \"Set-Cookie: wp_woocommerce_session_*************=*************; expires=Thu, 17-Sep-2020 06:21:23 GMT; Max-Age=172799; path=/; secure; HttpOnly\",
      \"Set-Cookie: woocommerce_items_in_cart=1; path=/\",
      \"Set-Cookie: woocommerce_cart_hash=*************; path=/\",
      \"Set-Cookie: wp_woocommerce_session_*************=*************; expires=Thu, 17-Sep-2020 06:21:23 GMT; Max-Age=172799; path=/; secure; HttpOnly\",
      \"Set-Cookie: cookielawinfo-checkbox-necessary=yes; expires=Tue, 15-Sep-2020 07:21:24 GMT; Max-Age=3600; path=/\",
      \"Set-Cookie: alg-wc-wl-user-id=*************; expires=Mon, 14-Dec-2020 06:21:24 GMT; Max-Age=7776000; path=/; secure\",
      \"Set-Cookie: alg-wc-wl-user-id=*************; expires=Mon, 14-Dec-2020 06:21:24 GMT; Max-Age=7776000; path=/; secure\",
      \"Set-Cookie: alg-wc-wl-user-id=*************; expires=Mon, 14-Dec-2020 06:21:24 GMT; Max-Age=7776000; path=/; secure\",
      \"Set-Cookie: alg-wc-wl-user-id=*************; expires=Mon, 14-Dec-2020 06:21:24 GMT; Max-Age=7776000; path=/; secure\",
      \"Set-Cookie: alg-wc-wl-user-id=*************; expires=Mon, 14-Dec-2020 06:21:24 GMT; Max-Age=7776000; path=/; secure\",
      \"Set-Cookie: alg-wc-wl-user-id=*************; expires=Mon, 14-Dec-2020 06:21:24 GMT; Max-Age=7776000; path=/; secure\",
      \"Set-Cookie: alg-wc-wl-user-id=*************; expires=Mon, 14-Dec-2020 06:21:24 GMT; Max-Age=7776000; path=/; secure\",
      \"Set-Cookie: alg-wc-wl-user-id=*************; expires=Mon, 14-Dec-2020 06:21:24 GMT; Max-Age=7776000; path=/; secure\",
      \"Set-Cookie: alg-wc-wl-user-id=*************; expires=Mon, 14-Dec-2020 06:21:24 GMT; Max-Age=7776000; path=/; secure\",
      \"Set-Cookie: alg-wc-wl-user-id=*************; expires=Mon, 14-Dec-2020 06:21:24 GMT; Max-Age=7776000; path=/; secure\",
      \"Set-Cookie: alg-wc-wl-user-id=*************; expires=Mon, 14-Dec-2020 06:21:24 GMT; Max-Age=7776000; path=/; secure\",
      \"Set-Cookie: alg-wc-wl-user-id=*************; expires=Mon, 14-Dec-2020 06:21:24 GMT; Max-Age=7776000; path=/; secure\",
      \"Set-Cookie: alg-wc-wl-user-id=*************; expires=Mon, 14-Dec-2020 06:21:24 GMT; Max-Age=7776000; path=/; secure\",
      \"Set-Cookie: alg-wc-wl-user-id=*************; expires=Mon, 14-Dec-2020 06:21:24 GMT; Max-Age=7776000; path=/; secure\",
      \"Set-Cookie: alg-wc-wl-user-id=*************; expires=Mon, 14-Dec-2020 06:21:24 GMT; Max-Age=7776000; path=/; secure\",
      \"Set-Cookie: alg-wc-wl-user-id=*************; expires=Mon, 14-Dec-2020 06:21:24 GMT; Max-Age=7776000; path=/; secure\",
      \"Set-Cookie: alg-wc-wl-user-id=*************; expires=Mon, 14-Dec-2020 06:21:24 GMT; Max-Age=7776000; path=/; secure\",
      \"Set-Cookie: alg-wc-wl-user-id=*************; expires=Mon, 14-Dec-2020 06:21:24 GMT; Max-Age=7776000; path=/; secure\",
      \"Set-Cookie: alg-wc-wl-user-id=*************; expires=Mon, 14-Dec-2020 06:21:24 GMT; Max-Age=7776000; path=/; secure\",
      \"Set-Cookie: alg-wc-wl-user-id=*************; expires=Mon, 14-Dec-2020 06:21:24 GMT; Max-Age=7776000; path=/; secure\",
      \"Set-Cookie: alg-wc-wl-user-id=*************; expires=Mon, 14-Dec-2020 06:21:24 GMT; Max-Age=7776000; path=/; secure\",
      \"Set-Cookie: alg-wc-wl-user-id=*************; expires=Mon, 14-Dec-2020 06:21:24 GMT; Max-Age=7776000; path=/; secure\",
      \"Set-Cookie: alg-wc-wl-user-id=*************; expires=Mon, 14-Dec-2020 06:21:24 GMT; Max-Age=7776000; path=/; secure\",
      \"Set-Cookie: alg-wc-wl-user-id=*************; expires=Mon, 14-Dec-2020 06:21:24 GMT; Max-Age=7776000; path=/; secure\",
      \"Set-Cookie: alg-wc-wl-user-id=*************; expires=Mon, 14-Dec-2020 06:21:24 GMT; Max-Age=7776000; path=/; secure\",
      \"Set-Cookie: alg-wc-wl-user-id=*************; expires=Mon, 14-Dec-2020 06:21:24 GMT; Max-Age=7776000; path=/; secure\",
      \"Set-Cookie: alg-wc-wl-user-id=*************; expires=Mon, 14-Dec-2020 06:21:24 GMT; Max-Age=7776000; path=/; secure\",
      \"Set-Cookie: alg-wc-wl-user-id=*************; expires=Mon, 14-Dec-2020 06:21:24 GMT; Max-Age=7776000; path=/; secure\",
      \"Set-Cookie: alg-wc-wl-user-id=*************; expires=Mon, 14-Dec-2020 06:21:24 GMT; Max-Age=7776000; path=/; secure\",
      \"Set-Cookie: alg-wc-wl-user-id=*************; expires=Mon, 14-Dec-2020 06:21:24 GMT; Max-Age=7776000; path=/; secure\",
      \"Set-Cookie: woocommerce_items_in_cart=1; path=/\",
      \"Set-Cookie: woocommerce_cart_hash=*************; path=/\",
      \"Set-Cookie: wp_woocommerce_session_*************; expires=Thu, 17-Sep-2020 06:21:23 GMT; Max-Age=172799; path=/; secure; HttpOnly\",
      \"Vary: Accept-Encoding\",
      \"Content-Encoding: gzip\",
      \"Content-Type: text/html; charset=UTF-8\",
      \"x-host: ******.com\",
      \"xkey: ******.com  ******.com/product-category/womens/\",
      \"X-Debug-Backend: web6_cst_webpod6__cph3_one_com\",
      \"X-Varnish: 5570577\",
      \"Age: 0\",
    },
  },
  vdc = 0x7f469ac5d028 {
    nxt = (nil),
    retval = 0,
  },
  vcl = {
    name = \"boot\",
    busy = 1034,
    discard = 0,
    state = auto,
    temp = warm,
    conf = {
      syntax = \"41\",
      srcname = {
        \"/etc/varnish/onecom.vcl\",
        \"Builtin\",
        \"/etc/varnish/vmods.vcl\",
        \"/etc/varnish/backends/backends.vcl\",
        \"/etc/varnish/backends/directors.vcl\",
        \"/etc/varnish/core/error.vcl\",
        \"/etc/varnish/core/core.vcl\",
        \"/etc/varnish/core/modules.vcl\",
        \"/etc/varnish/core/modules/purge.vcl\",
        \"/etc/varnish/core/modules/security.vcl\",
        \"/etc/varnish/core/modules/mailconfig.vcl\",
        \"/etc/varnish/core/modules/remove_cookies.vcl\",
        \"/etc/varnish/core/modules/debug.vcl\",
        \"/etc/varnish/core/modules/devicedetect.vcl\",
        \"/etc/varnish/core/modules/subs.vcl\",
        \"/etc/varnish/app.vcl\",
        \"/etc/varnish/apps/apps_recv.vcl\",
        \"/etc/varnish/apps_status.vcl\",
        \"/etc/varnish/apps/cookies/recv.vcl\",
        \"/etc/varnish/apps_status.vcl\",
        \"/etc/varnish/apps_status.vcl\",
        \"/etc/varnish/apps/diagnostics/recv.vcl\",
        \"/etc/varnish/apps_status.vcl\",
        \"/etc/varnish/apps/gallery/recv.vcl\",
        \"/etc/varnish/apps_status.vcl\",
        \"/etc/varnish/apps_status.vcl\",
        \"/etc/varnish/apps_status.vcl\",
        \"/etc/varnish/apps/impro/recv.vcl\",
        \"/etc/varnish/apps_status.vcl\",
        \"/etc/varnish/apps_status.vcl\",
        \"/etc/varnish/apps_status.vcl\",
        \"/etc/varnish/apps/webcluster/recv.vcl\",
        \"/etc/varnish/customer_rules/customer_recv.vcl\",
        \"/etc/varnish/apps/apps_synth.vcl\",
        \"/etc/varnish/apps/cookies/synth.vcl\",
        \"/etc/varnish/apps/apps_backend_response.vcl\",
        \"/etc/varnish/apps/webcluster/backend_response.vcl\",
        \"/etc/varnish/customer_rules/customer_backend_response.vcl\",
        \"/etc/varnish/apps/apps_backend_error.vcl\",
        \"/etc/varnish/apps/webcluster/backend_error.vcl\",
        \"/etc/varnish/apps/apps_hash.vcl\",
        \"/etc/varnish/apps/webcluster/hash.vcl\",
        \"/etc/varnish/apps/apps_deliver.vcl\",
        \"/etc/varnish/apps/webcluster/deliver.vcl\",
        \"/etc/varnish/apps/apps_miss.vcl\",
        \"/etc/varnish/apps/webcluster/miss.vcl\",
        \"/etc/varnish/customer_rules/customer_misspass.vcl\",
        \"/etc/varnish/apps/apps_pass.vcl\",
        \"/etc/varnish/apps/webcluster/pass.vcl\",
        \"/etc/varnish/customer_rules/customer_misspass.vcl\",
      instances = {
        \"autodiscover\" = 0x7f46da20c1e0,
        \"diagnostics\" = 0x7f46da20c2c0,
        \"formmail\" = 0x7f46da20c160,
        \"gallery\" = 0x7f46da20c220,
        \"impro\" = 0x7f46da20c120,
        \"webcluster\" = 0x7f46da20c280,
        \"webforward\" = 0x7f46da20c300,
        \"webshop\" = 0x7f46da20c1a0,
      },
    },
  },
  objcore[REQ] = 0x7f4677891180 {
    refcnt = 3,
    flags = {hfm},
    exp_flags = {refd},
    boc = 0x7f4690999080 {
      refcnt = 1,
      state = stream,
      vary = (nil),
      stevedore_priv = (nil),
    },
    exp = {1600150884.796366, 0.000000, 10.000000, 0.000000},
    objhead = 0x7f4689bb4390,
    stevedore = 0x7f46da0db300 (malloc Transient) {
      Simple = 0x7f469d3a6000,
      Obj = 0x7f468c5096f8 {priv=0x7f468c5096f0, ptr=0x7f469d3a6000, len=6467, space=6472},
      LEN = 0x0...0,
      VXID = 0x01120007,
      FLAGS = 0x02,
      GZIPBITS = 0x0000000000000050000000000000000000000000000000000000000000000000,
      LASTMODIFIED = 0x41d7d81759000000,
      VARY = {len=27, ptr=0x7f469d3a6078},
      HEADERS = {len=6320, ptr=0x7f469d3a6093},
      Body = 0x7f469cb431f8 {priv=0x7f469cb431f0, ptr=0x7f467251b000, len=11430, space=16384},
    },
  },
  flags = {
  },
  privs = 0x7f469ac5b200 {
    priv {p (nil) l 0 f (nil)} vmod 7f46b16614a8
  },
  top = 0x7f469ac5d0b0 {
    req = 0x7f469ac5b020 {
      [Already dumped, see above]
    },
    privs = 0x7f469ac5d0c8 {
    },
  },
},
thr.busyobj = (nil) {
},
vmods = {
  std = {0x7f46da10c0f0, Varnish trunk 7786bd8d1396f3a60c23057091c5b34b81e3e275, 0.0},
  one = {0x7f46da10c1d0, Varnish trunk 7786bd8d1396f3a60c23057091c5b34b81e3e275, 0.0},
  maxminddb = {0x7f46da10c240, Varnish trunk 7786bd8d1396f3a60c23057091c5b34b81e3e275, 0.0},
  directors = {0x7f46da10c2b0, Varnish trunk 7786bd8d1396f3a60c23057091c5b34b81e3e275, 0.0},
  xkey = {0x7f46da10c390, Varnish trunk 7786bd8d1396f3a60c23057091c5b34b81e3e275, 0.0},
  header = {0x7f46da10c400, Varnish trunk 7786bd8d1396f3a60c23057091c5b34b81e3e275, 11.0},
  cookie = {0x7f46da10c470, Varnish trunk 7786bd8d1396f3a60c23057091c5b34b81e3e275, 0.0},
  vsthrottle = {0x7f46da10c4e0, Varnish trunk 7786bd8d1396f3a60c23057091c5b34b81e3e275, 11.0},
  tcp = {0x7f46da10c550, Varnish trunk 7786bd8d1396f3a60c23057091c5b34b81e3e275, 0.0},
  bodyaccess = {0x7f46da10c5c0, Varnish trunk 7786bd8d1396f3a60c23057091c5b34b81e3e275, 0.0},
  blob = {0x7f46da10c630, Varnish trunk 7786bd8d1396f3a60c23057091c5b34b81e3e275, 0.0},
  digest = {0x7f46da10c6a0, Varnish trunk 7786bd8d1396f3a60c23057091c5b34b81e3e275, 0.0},
},

```

This error went from "Error          workspace_client overflow to" to panic from 4dc4d4564c1b6a87dbb311616cc8edc4b104c35f to 7786bd8d1396f3a60c23057091c5b34b81e3e275

nm -an /usr/sbin/varnishd: [nm_7786bd8d1396f3a60c23057091c5b34b81e3e275.txt](https://github.com/varnishcache/varnish-cache/files/5222871/nm_7786bd8d1396f3a60c23057091c5b34b81e3e275.txt)

**Comment by Dridi** [comment]
So we run into a failure situation where the workspace did not overflow prior to entering this function.

I say we remove the assertion and restore the previous mark-overflow, and later try to understand why the workspace was not marked as overflowed.

**Comment by Dridi** [comment]
This one almost slipped through the cracks and landed via #3342, thanks a lot!

It was not initially part of my changes, but it came up during the review (https://github.com/varnishcache/varnish-cache/pull/3342#discussion_r453553955) and I failed to identify that we could very well not have technically overflowed the workspace itself, and only the portion allocated to keep track of headers.
